File: leetcode/src/main/java/leetcode/design/LRUCache.java
package leetcode.design;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

/**
 * https://leetcode-cn.com/problems/lru-cache-lcci/
 *
 * medium
 */
class LRUCache {

    private class DLinkedNode {
        int key;
        int value;
        DLinkedNode prev;
        DLinkedNode next;
        public DLinkedNode() {}
        public DLinkedNode(int _key, int _value) {key = _key; value = _value;}
    }

    private DLinkedNode head;
    private DLinkedNode tail;
    private Map<Integer/* key */, DLinkedNode> map;
    private int capacity;
    private int size;


    public LRUCache(int capacity) {
        if (capacity <= 0) {
            throw new RuntimeException("capacity > 0");
        }
        this.size = 0;
        this.capacity = capacity;
        map = new HashMap<>();
        head = new DLinkedNode();
        tail = new DLinkedNode();
        head.next = tail;
        tail.prev = head;
    }

    public int get(int key) {
        DLinkedNode node = map.get(key);
        if (node == null) {
            return -1;
        }
        moveToTop(node);
        return node.value;
    }

    public void put(int key, int value) {
        // 如果已经在map里，移动到头位
        DLinkedNode node = map.get(key);
        if (node != null) {
            node.value = value;
            moveToTop(node);
        // 如果不在map里，新增一个node，移动到头位
        } else {
            node = new DLinkedNode(key, value);
            this.size++;
            addToTop(node);
            map.put(key, node);
        }

        // 判断capacity，回收
        if (size > capacity) {
            DLinkedNode last = tail.prev;
            remove(last);
            map.remove(last.key);
            size--;
        }
    }

    private void addToTop(DLinkedNode node) {
        DLinkedNode next = head.next;

        next.prev = node;
        head.next = node;
        node.next = next;
        node.prev = head;
    }

    private void moveToTop(DLinkedNode node) {
        remove(node);
        addToTop(node);
    }

    private void remove(DLinkedNode node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }
}

File: go_251215/LRUCache.go
package main

type LRUCache struct {
	head *node
	tail *node
	cap  int
	size int
	data map[int]*node
}

type node struct {
	val  int
	key  int
	prev *node
	next *node
}

func Constructor(capacity int) LRUCache {
	if capacity <= 0 {
		panic("cap must > 0")
	}
	head, tail := &node{}, &node{}
	head.next = tail
	tail.prev = head

	return LRUCache{cap: capacity, head: head, tail: tail, data: map[int]*node{}}
}

func (c *LRUCache) Get(key int) int {
	n, ok := c.data[key]
	if !ok {
		return -1
	}
	c.moveTo(n, c.head)
	return n.val
}

func (c *LRUCache) moveTo(n *node, newPrev *node) {
	if n.prev != nil {
		n.prev.next = n.next
	}
	if n.next != nil {
		n.next.prev = n.prev
	}

	beforeTop := newPrev.next
	newPrev.next = n
	n.next = beforeTop
	n.prev = newPrev
	if beforeTop != nil {
		beforeTop.prev = n
	}
}

func (c *LRUCache) remove(n *node) {
	n.prev.next = n.next
	n.next.prev = n.prev
	n.prev = nil
	n.next = nil
}

func (c *LRUCache) Put(key int, value int) {
	// 1) key 已存在：更新 + 提到头部
	if n, ok := c.data[key]; ok {
		n.val = value
		c.moveTo(n, c.head)
		return
	}

	// 2) 新 key：创建 + 插入
	n := &node{val: value, key: key}
	c.moveTo(n, c.head)
	c.data[key] = n
	c.size++

	// 3) 超容量：淘汰尾部
	if c.size > c.cap {
		last := c.tail.prev
		delete(c.data, last.key)
		c.remove(last)
		c.size--
	}
}
